### PR TITLE
fix(cypress): preserve lifecycle finalization errors

### PR DESCRIPTION
**Pull request by juan-fernandez**
### What does this PR do?
Ensures Cypress Test Optimization finalization still runs when a user `after:spec` or `after:run` handler throws or rejects.

Datadog now finalizes the test session trace as failed, preserves completed test events, and then propagates the original user error back to Cypress. Finalization remains bounded when screenshot uploads or writer requests are pending. The same behavior works when auto-instrumentation encounters a manual Cypress plugin from an older dd-trace copy.

### Motivation
A terminal Cypress hook can prevent the next lifecycle event from running. For example:

```js
module.exports = {
  e2e: {
    setupNodeEvents (on) {
      on('after:spec', async () => {
        throw new Error('failed to clean up test data')
      })
    }
  }
}
```

Previously, this rejection could stop Datadog's `after:spec` finalizer and prevent Cypress from reaching `after:run`. Completed tests might already have been submitted, while their suite, module, or session events were missing—or the test session trace was reported as passing even though Cypress exited non-zero.

With this PR, Datadog runs its finalizer after the user handler, marks the suite, module, and session as failed with `failed to clean up test data`, submits the complete test session trace within the existing deadline, and rethrows that same error so Cypress keeps its normal non-zero exit behavior.

### Additional Notes
Builds on #9790's deferred test session trace support. This PR intentionally excludes `experimentalInteractiveRunEvents`; that behavior remains isolated in the later interactive-runs PR.

Validation:
- Focused Cypress lifecycle and configuration regressions
- Real Cypress CommonJS and ESM integration cases
- Cypress ITR integration suite
- Modified-file linting; full lint is currently blocked locally by the missing OpenFeature vendor bundle

### CODE REVIEW

**Comment by juan-fernandez** [comment]
@codex review

File: packages/datadog-plugin-cypress/src/cypress-plugin.js [hunk 1]
```diff
@@ -1608,16 +1646,23 @@ class CypressPlugin {
 
     finishSuite()
 
+    if (error) {
+      this.abortPendingScreenshotUploads(error)
+      return this.afterRun(undefined, error)
```
[the remark this answers is not in the document]
Comment (juan-fernandez) [reply packages/datadog-plugin-cypress/src/cypress-plugin.js]: Fixed in fix(cypress): reset interactive run state safely (https://github.com/DataDog/dd-trace-js/commit/41fb6237a).

File: packages/datadog-plugin-cypress/src/cypress-plugin.js [hunk 2]
```diff
@@ -1063,9 +1079,13 @@ class CypressPlugin {
...
-    // We need to make sure that the plugin is initialized before running the tests
-    // This is for the case where the user has not returned the promise from the init function
-    await this.libraryConfigurationPromise
+    if (this._isInit) {
+      // The user may not have returned the promise from the init function.
+      await this.libraryConfigurationPromise
+    } else {
+      // Cypress open reuses the same plugin process for every interactive run.
+      await this.init(this.tracer, this.cypressConfig)
+    }
```
[the remark this answers is not in the document]
Comment (juan-fernandez) [reply packages/datadog-plugin-cypress/src/cypress-plugin.js:1099]: Fixed in fix(cypress): reset interactive run state safely (https://github.com/DataDog/dd-trace-js/commit/41fb6237a).

File: packages/datadog-plugin-cypress/src/cypress-plugin.js [hunk 3]
```diff
@@ -1572,27 +1604,33 @@ class CypressPlugin {
...
+        const status = error ? 'fail' : getSuiteStatus(stats)
         this.testSuiteSpan.setTag(TEST_STATUS, status)
 
-        if (latestError) {
-          this.testSuiteSpan.setTag('error', latestError)
+        if (error || latestError) {
+          this.testSuiteSpan.setTag('error', error || latestError)
         }
+        const exporter = this.tracer._tracer._exporter
+        if (exporter.deferTestSuiteSpan) exporter.deferTestSuiteSpan(this.testSuiteSpan)
```
[the remark this answers is not in the document]
Comment (juan-fernandez) [reply packages/datadog-plugin-cypress/src/cypress-plugin.js]: Fixed in fix(cypress): reset interactive run state safely (https://github.com/DataDog/dd-trace-js/commit/41fb6237a).

**Comment by juan-fernandez** [comment]
@codex review

File: packages/datadog-instrumentations/src/cypress-config.js [hunk 4]
```diff
@@ -433,13 +487,20 @@ function registerDdTraceHooks (
     if (generatedSupportFiles) cleanupGeneratedFiles(generatedSupportFiles)
   }
 
-  const registerAfterRunWithCleanup = () => {
+  const registerAfterRunWithCleanup = (datadogHandler) => {
+    const handlers = datadogHandler
+      ? userAfterRunHandlers.filter(handler => handler !== datadogHandler)
```
[the remark this answers is not in the document]
Comment (juan-fernandez) [reply packages/datadog-instrumentations/src/cypress-config.js:520]: Fixed in fix(cypress): coordinate legacy finalizers (https://github.com/DataDog/dd-trace-js/commit/fb7ec9d9ffc885b4479231aba776e7da5e75d24d).

File: packages/datadog-instrumentations/src/cypress-config.js [hunk 5]
```diff
@@ -387,21 +405,57 @@ function registerManualAfterScreenshotHandlers (on, handlers, datadogHandler) {
...
+ *
+ * @param {Promise<void>} userHandlers collected user-handler chain
+ * @param {(userError?: unknown) => unknown} finalizer Datadog finalizer
+ * @returns {Promise<unknown>} finalizer result
+ */
+function finalizeAfterUserHandlers (userHandlers, finalizer) {
+  return userHandlers.then(
+    () => finalizer(),
+    userError => Promise.resolve().then(() => {
+      const finalizerError = userError || new Error('Cypress user handler rejected without an error')
```
[the remark this answers is not in the document]
Comment (juan-fernandez) [reply packages/datadog-instrumentations/src/cypress-config.js]: Fixed in fix(cypress): coordinate legacy finalizers (https://github.com/DataDog/dd-trace-js/commit/fb7ec9d9ffc885b4479231aba776e7da5e75d24d).

File: packages/datadog-instrumentations/src/cypress-config.js [hunk 6]
```diff
@@ -497,14 +564,18 @@ function wrapSetupNodeEvents (originalSetupNodeEvents) {
...
       afterScreenshotHandler: undefined,
     }
 
     const wrappedOn = (event, handler) => {
       if (event === 'after:spec') {
         userAfterSpecHandlers.push(handler)
+        if (isDatadogAfterSpecHandler(handler)) manualPlugin.afterSpecHandler = handler
       } else if (event === 'after:run') {
         userAfterRunHandlers.push(handler)
+        if (isDatadogAfterRunHandler(handler)) manualPlugin.afterRunHandler = handler
```
[the remark this answers is not in the document]
Comment (juan-fernandez) [reply packages/datadog-instrumentations/src/cypress-config.js:628]: Addressed in [fix(cypress): support older manual finalizers](https://github.com/DataDog/dd-trace-js/commit/fa458972d5fcf48fb1da0ac7fe025258434198e2). The task-registration fallback now captures unmarked legacy after:spec and after:run handlers.

File: packages/datadog-instrumentations/src/cypress-config.js [hunk 7]
```diff
@@ -451,9 +518,9 @@ function registerDdTraceHooks (
   }
 
   if (manualPlugin.detected) {
-    registerAfterSpecHandlers(on, userAfterSpecHandlers)
+    registerAfterSpecHandlers(on, userAfterSpecHandlers, manualPlugin.afterSpecHandler)
```
[the remark this answers is not in the document]
Comment (juan-fernandez) [reply packages/datadog-instrumentations/src/cypress-config.js]: Addressed in [fix(cypress): support older manual finalizers](https://github.com/DataDog/dd-trace-js/commit/fa458972d5fcf48fb1da0ac7fe025258434198e2). Manual after:spec failures now remove generated support files even when Cypress suppresses after:run.

**Comment by juan-fernandez** [comment]
@codex review

File: packages/datadog-instrumentations/src/cypress-config.js [hunk 8]
```diff
@@ -387,21 +406,74 @@ function registerManualAfterScreenshotHandlers (on, handlers, datadogHandler) {
...
+  on('after:spec', (spec, results) => {
+    const callHandler = datadogHandler
+      ? handler => runUserHandler(() => handler(spec, results))
+      : handler => handler(spec, results)
+    const chain = userHandlers.reduce(
+      (promise, handler) => promise.then(() => callHandler(handler)),
+      Promise.resolve()
+    )
+    if (!datadogHandler) return chain
+    const finalization = finalizeAfterUserHandlers(chain, (...args) => datadogHandler(spec, results, ...args))
```
[the remark this answers is not in the document]
Comment (juan-fernandez) [reply packages/datadog-instrumentations/src/cypress-config.js:478]: Addressed in [fix(cypress): adapt legacy manual finalizers](https://github.com/DataDog/dd-trace-js/commit/bb8ea1a1636ac9fcb2ec9f030cf1cc4321d5b873). When an unmarked pre-change manual copy is detected, the current auto-instrumented copy now owns terminal finalization and Datadog tasks; the legacy terminal handlers are excluded. The real cross-copy regression verifies one failed test session trace with the original after:spec error in CommonJS, ESM, and the v5 package configuration.

File: packages/datadog-instrumentations/src/cypress-config.js [hunk 9]
```diff
@@ -433,13 +505,23 @@ function registerDdTraceHooks (
...
+  const registerAfterRunWithCleanup = (datadogHandler) => {
+    const handlers = datadogHandler
+      ? userAfterRunHandlers.filter(handler => handler !== datadogHandler)
+      : userAfterRunHandlers
+
     on('after:run', (results) => {
-      const chain = userAfterRunHandlers.reduce(
-        (p, h) => p.then(() => h(results)),
+      const callHandler = datadogHandler
+        ? handler => runUserHandler(() => handler(results))
```
[the remark this answers is not in the document]
Comment (juan-fernandez) [reply packages/datadog-instrumentations/src/cypress-config.js:525]: Fixed in [fix(cypress): defer older lifecycle helpers](https://github.com/DataDog/dd-trace-js/commit/c9dacc2b785f0bf3657d536103b93b38c4dcb1ec). The newer Cypress instrumentation now wraps the older copy’s direct `after:run` and `after:spec` helper implementations when Cypress activates, so they honor the shared deferral context even though the old helpers do not implement it. The regression uses genuine pre-deferral helper code together with unmarked manual-plugin handlers and verifies one failed test session trace, the original error, the completed test event, and a non-zero exit in both CommonJS and ESM.

File: packages/datadog-instrumentations/src/cypress-config.js [hunk 10]
```diff
@@ -450,13 +541,23 @@ function registerDdTraceHooks (
     on('task', noopTask)
   }
 
-  if (manualPlugin.detected) {
-    registerAfterSpecHandlers(on, userAfterSpecHandlers)
+  if (manualPlugin.detected &&
+    (supportsErrorAwareFinalization(manualPlugin) || !setupNodeEventsCh.hasSubscribers)) {
```
[the remark this answers is not in the document]
Comment (juan-fernandez) [reply packages/datadog-instrumentations/src/cypress-config.js]: Fixed in [fix(cypress): preserve manual plugin ownership](https://github.com/DataDog/dd-trace-js/commit/ebb502aea4a159bd07941bd159a8759d3c400383). The real manual task map now carries a copy-local owner token. Wrapped lifecycle callbacks can lose their symbols without losing same-copy ownership, while a different dd-trace copy cannot be mistaken for the active singleton. Added unit and real CommonJS/ESM coverage proving tasks and the complete test session trace remain available.

File: packages/datadog-instrumentations/src/cypress-config.js [hunk 11]
```diff
@@ -387,21 +415,74 @@ function registerManualAfterScreenshotHandlers (on, handlers, datadogHandler) {
...
+ *
+ * @param {Promise<void>} userHandlers collected user-handler chain
+ * @param {(userError?: unknown) => unknown} finalizer Datadog finalizer
+ * @returns {Promise<unknown>} finalizer result
+ */
+function finalizeAfterUserHandlers (userHandlers, finalizer) {
+  return userHandlers.then(
+    () => finalizer(),
+    userError => Promise.resolve().then(() => {
+      return finalizer(normalizeUserHandlerError(userError))
```
[the remark this answers is not in the document]
Comment (juan-fernandez) [reply packages/datadog-instrumentations/src/cypress-config.js]: Fixed in [fix(cypress): preserve manual plugin ownership](https://github.com/DataDog/dd-trace-js/commit/ebb502aea4a159bd07941bd159a8759d3c400383). `finalizeAfterUserHandlers` now lives in `finalization.js` and both the manual-plugin subscriber and config instrumentation import the same promise ordering, error precedence, and finalizer-error logging policy.

File: packages/datadog-instrumentations/src/cypress-config.js [hunk 12]
```diff
@@ -497,22 +598,31 @@ function wrapSetupNodeEvents (originalSetupNodeEvents) {
...
+        if (isDatadogAfterRunHandler(handler)) manualPlugin.afterRunHandler = handler
       } else if (event === 'after:screenshot') {
         userAfterScreenshotHandlers.push(handler)
       } else {
         if (event === 'task' && isDatadogTaskRegistration(handler)) {
           manualPlugin.detected = true
-          manualPlugin.afterScreenshotHandler = userAfterScreenshotHandlers.at(-1)
+          manualPlugin.afterSpecHandler ||= userAfterSpecHandlers.at(-1)
+          manualPlugin.afterRunHandler ||= userAfterRunHandlers.at(-1)
+          manualPlugin.afterScreenshotHandler ||= userAfterScreenshotHandlers.at(-1)
```
[the remark this answers is not in the document]
Comment (juan-fernandez) [reply packages/datadog-instrumentations/src/cypress-config.js]: Fixed in [fix(cypress): preserve manual plugin ownership](https://github.com/DataDog/dd-trace-js/commit/ebb502aea4a159bd07941bd159a8759d3c400383). The older-plugin fallback now claims handlers only when it observes the contiguous manual lifecycle registration sequence before the Datadog task map. A no-op manual plugin, which registers only tasks, leaves all preceding user `after:spec`, `after:run`, and `after:screenshot` handlers intact. Added a focused regression for that path.

File: packages/datadog-instrumentations/src/cypress-config.js [hunk 13]
```diff
@@ -450,13 +541,23 @@ function registerDdTraceHooks (
...
     registerManualAfterScreenshotHandlers(on, userAfterScreenshotHandlers, manualPlugin.afterScreenshotHandler)
-    registerAfterRunWithCleanup()
+    registerAfterRunWithCleanup(manualPlugin.afterRunHandler)
+    on('task', manualPlugin.taskHandler)
     return config
   }
 
+  if (manualPlugin.detected) {
+    userAfterSpecHandlers = userAfterSpecHandlers.filter(handler => handler !== manualPlugin.afterSpecHandler)
+    userAfterRunHandlers = userAfterRunHandlers.filter(handler => handler !== manualPlugin.afterRunHandler)
```
[the remark this answers is not in the document]
Comment (juan-fernandez) [reply packages/datadog-instrumentations/src/cypress-config.js:561]: Fixed in [fix(cypress): isolate lifecycle finalization failures](https://github.com/DataDog/dd-trace-js/commit/add2e9147f1a62ccb0f63ca69db35f64f170e59f). The config wrapper now defers before:run registration until it knows which dd-trace copy owns the lifecycle, and filters the older copy's before:run handler during the ownership handoff. The regression verifies that only the current handler is registered.

File: packages/datadog-plugin-cypress/src/cypress-plugin.js [hunk 14]
```diff
@@ -1252,18 +1283,29 @@ class CypressPlugin {
...
       const hasBackfilledCoverage = this.applySkippedCoverageToTestSessionCoverage()
       const testCodeCoverageLinesTotal = this.getTestCodeCoverageLinesTotal(hasBackfilledCoverage)
 
       this.testModuleSpan.setTag(TEST_STATUS, testStatus)
       this.testSessionSpan.setTag(TEST_STATUS, testStatus)
+      for (const span of this.finishedTestSuiteSpans) {
+        if (error) {
+          span.setTag(TEST_STATUS, 'fail')
+          span.setTag('error', error)
+        }
```
[the remark this answers is not in the document]
Comment (juan-fernandez) [reply packages/datadog-plugin-cypress/src/cypress-plugin.js:1302]: Fixed in [fix(cypress): isolate lifecycle finalization failures](https://github.com/DataDog/dd-trace-js/commit/add2e9147f1a62ccb0f63ca69db35f64f170e59f). after:spec errors now finalize the module/session without applying the error to previously completed suites. The new two-spec integration regression verifies the earlier suite stays pass while the current suite, module, and session carry the original error.

**Comment by juan-fernandez** [comment]
@codex review

File: packages/datadog-instrumentations/src/cypress-config.js [hunk 15]
```diff
@@ -492,28 +591,53 @@ function registerDdTraceHooks (
...
       } else if (event === 'after:screenshot') {
         userAfterScreenshotHandlers.push(handler)
       } else {
         if (event === 'task' && isDatadogTaskRegistration(handler)) {
           manualPlugin.detected = true
-          manualPlugin.afterScreenshotHandler = userAfterScreenshotHandlers.at(-1)
+          manualPlugin.ownsCurrentPlugin = isCurrentDatadogTaskRegistration(handler)
+          const [beforeRun, afterScreenshot, afterSpec, afterRun] = recentRegistrations
+          if (beforeRun?.event === 'before:run' &&
+            afterScreenshot?.event === 'after:screenshot' &&
```
[the remark this answers is not in the document]
Comment (juan-fernandez) [reply packages/datadog-instrumentations/src/cypress-config.js]: Fixed in [fix(cypress): support pre-screenshot manual plugins](https://github.com/DataDog/dd-trace-js/commit/5815607ac29b6097c88a52a4b12c1cc2a35de43a). Ownership detection now accepts both historical suffixes: before:run/after:spec/after:run/task and the newer shape with after:screenshot. The mixed-copy integration regression drops the screenshot registration, preserves the later user error, and verifies one failed test session trace. I also made the multi-spec suite test reject the second spec independent of Cypress's version-specific ordering, fixing the Cypress 12 CI failures.

**Comment by juan-fernandez** [comment]
@codex review

File: packages/datadog-plugin-cypress/src/cypress-plugin.js [hunk 16]
```diff
@@ -1572,27 +1615,33 @@ class CypressPlugin {
...
 
-        if (latestError) {
-          this.testSuiteSpan.setTag('error', latestError)
+        if (error || latestError) {
+          this.testSuiteSpan.setTag('error', error || latestError)
         }
+        const exporter = this.tracer._tracer._exporter
+        if (exporter?.deferTestSuiteSpan) exporter.deferTestSuiteSpan(this.testSuiteSpan)
         this.testSuiteSpan.finish()
+        this.finishedTestSuiteSpans.push(this.testSuiteSpan)
```
[the remark this answers is not in the document]
Comment (juan-fernandez) [reply packages/datadog-plugin-cypress/src/cypress-plugin.js]: Fixed in [fix(cypress): release suites without run finalization](https://github.com/DataDog/dd-trace-js/commit/26a7e4095). Completed suites are now deferred and retained only when Cypress has an `after:run` boundary: terminal runs, or interactive runs with `experimentalInteractiveRunEvents`. In ordinary `cypress open`, the suite follows normal completion and is not accumulated for a callback Cypress will never invoke. Regression coverage pins all three configurations.

**Comment by juan-fernandez** [comment]
@codex review

File: packages/datadog-instrumentations/src/cypress-config.js [hunk 17]
```diff
@@ -492,28 +591,56 @@ function registerDdTraceHooks (
...
-          manualPlugin.afterScreenshotHandler = userAfterScreenshotHandlers.at(-1)
+          manualPlugin.ownsCurrentPlugin = isCurrentDatadogTaskRegistration(handler)
+          const afterRun = recentRegistrations.at(-1)
+          const afterSpec = recentRegistrations.at(-2)
+          const possibleAfterScreenshot = recentRegistrations.at(-3)
+          const hasAfterScreenshot = possibleAfterScreenshot?.event === 'after:screenshot'
+          const beforeRun = hasAfterScreenshot ? recentRegistrations.at(-4) : possibleAfterScreenshot
+          if (beforeRun?.event === 'before:run' &&
+            afterSpec?.event === 'after:spec' &&
+            afterRun?.event === 'after:run') {
```
[the remark this answers is not in the document]
Comment (juan-fernandez) [reply packages/datadog-instrumentations/src/cypress-config.js]: Fixed in [fix(cypress): preserve hooks with no-op plugin](https://github.com/DataDog/dd-trace-js/commit/99ee4cb68). The manual plugin now marks its fallback task map explicitly. Config instrumentation recognizes that marker, does not claim adjacent lifecycle handlers, re-registers all user hooks, reuses the existing no-op tasks, and skips plugin activation. The regression now places before:run/after:spec/after:run directly next to the no-op task registration and verifies all user callbacks. The real invalid-initialization path passes in CJS and ESM.

**Comment by juan-fernandez** [comment]
@codex review

**Review by anmarchenko — APPROVED** [review]
